6445283: ProgressMonitorInputStream not large file aware (>2GB)

prsadhuk opened the pull request:
If ProgressMonitor has to show progress for reading file > 2GB, it goes to 100% and then again start from 0. 
This is because
it uses "int" to store bytes read (`nread`) and when it reads data from file, it adds 
"number-bytes-to-read "nr" to number-bytes-already-read "nread" variable [`nread += nr`] which can cause it to overflow and so "nread" becomes -ve.

Fix is to check if adding bytes-to-read to number-bytes-already-read will exceed MAX_INT, then set Progress to max so that ProgressMonitor can close the tracker 
https://github.com/openjdk/jdk/blob/master/src/java.desktop/share/classes/javax/swing/ProgressMonitor.java#L265

No regression test is added as it involves using filesize of >2GB.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-6445283](https://bugs.openjdk.org/browse/JDK-6445283): ProgressMonitorInputStream not large file aware (>2GB)


### Reviewers
 * [Alexander Zvegintsev](https://openjdk.org/census#azvegint) (@azvegint - **Reviewer**) ⚠️ Review applies to [a34eae0b](https://git.openjdk.org/jdk/pull/9588/files/a34eae0b84af452ca5285c6564ffe9cf189e5659)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9588/head:pull/9588` \
`$ git checkout pull/9588`

Update a local copy of the PR: \
`$ git checkout pull/9588` \
`$ git pull https://git.openjdk.org/jdk pull/9588/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9588`

View PR using the GUI difftool: \
`$ git pr show -t 9588`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9588.diff">https://git.openjdk.org/jdk/pull/9588.diff</a>

</details>

azvegint commented on src/java.desktop/share/classes/javax/swing/ProgressMonitorInputStream.java:
```diff
@@ -145,7 +151,13 @@ public int read(byte[] b,
+            } else {
+                monitor.setProgress(monitor.getMaximum());
+            }
+        }
```
Looks like Lines 154-167 duplicates code from lines 129-142.
Can we consider to extract it so another method?

azvegint commented:
> Fix is to check if adding bytes-to-read to number-bytes-already-read will exceed MAX_INT, then set Progress to max so that ProgressMonitor can close the tracker
https://github.com/openjdk/jdk/blob/master/src/java.desktop/share/classes/javax/swing/ProgressMonitor.java#L265

After applying the fix, I see that the tracker window closes when it reaches the limit, while the progress may be still ongoing for a prolonged period time. It doesn't look right to me.

I think we can either change this close condition, either fake progress to `MAX_INT-1` until we actually reach the end.

Anyway it seems like a workaround, real solution would be to migrate to long(probably by adding new constructors and methods to keep compatibility).


> No regression test is added as it involves using filesize of >2GB.

It doesn't have to be a FileInputStream, since ProgressMonitorInputStream accepts InputStream. So we can make a dummy one with no real data backing it.

prsadhuk commented:
> After applying the fix, I see that the tracker window closes when it reaches the limit, while the progress may be still ongoing for a prolonged period time. It doesn't look right to me.
> 
> I think we can either change this close condition, either fake progress to MAX_INT-1 until we actually reach the end.
> 
> Anyway it seems like a workaround, real solution would be to migrate to long(probably by adding new constructors and methods to keep compatibility).

Yes, the tracker closes when it reaches MAX INT limit 2147483647 since the progress tracking method setProgress() accepts "int". 
We can keep the progress meter at end ie 100% without closing even after reading 2147483647  bytes, till it reaches EOF
or 
we could add new setProgress() method with long as you suggested. I had thought about it but was not sure about the practicality of this scenario, so just thought of addressing by closing the tracker.
Also, InputStream.available() returns an "int" so I guessed filesize > MAX_INT is not supported which is another reason I did not go for "long"

I will update the fix for the former.

> It doesn't have to be a FileInputStream, since ProgressMonitorInputStream accepts InputStream. So we can make a dummy one with no real data backing it.

I tried creating ByteArrayInputStream of 3GB but it fails with NegativeArraySizeException because of overflow so I am not sure if it's possible.

azvegint commented:
> Yes, the tracker closes when it reaches MAX INT limit 2147483647 since the progress tracking method setProgress() accepts "int". We can keep the progress meter at end ie 100% without closing even after reading 2147483647 bytes, till it reaches EOF or we could add new setProgress() method with long as you suggested. I had thought about it but was not sure about the practicality of this scenario, so just thought of addressing by closing the tracker. Also, InputStream.available() returns an "int" so I guessed filesize > MAX_INT is not supported which is another reason I did not go for "long"

https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/InputStream.html#available()

>Returns an estimate of the number of bytes that can be read (or skipped over) from this input stream without blocking, which may be 0, or 0 when end of stream is detected. 


So `InputStream.available()` is not intended to get a real input stream data length.

Maybe we should work with `InputStream.available()` not only in constructor, but in `ProgressMonitorInputStream.read()` methods too, since `available()` value can change during the data read.

For example, we have `FileInputStream` with a 4GB file:

- at the beginning we have MAX INT `available()`
- after 1GB read we still have MAX INT available
but 
- after 3GB read we have 1GB available
- after 4GB read we have 0 available

And we can somehow update progress bar based on this data.

prsadhuk commented:
But it seems in.available() does not change when we read from the stream. For eg, if we read 512 bytes from stream, I would expect (2147483647 -512) bytes but I still see available() returns 2147483647 so this will not work AFICS

azvegint commented:
> But it seems in.available() does not change when we read from the stream. For eg, if we read 512 bytes from stream, I would expect (2147483647 -512) bytes but I still see available() returns 2147483647 so this will not work AFICS

It does, you can play with [this](https://gist.github.com/azvegint/65b30efa2749e5af3fae6a247c344677) and provide large file as argument.

However, frequent calling of `available()` may reduce performance, so I agree with your solution.


> I tried creating ByteArrayInputStream of 3GB but it fails with NegativeArraySizeException because of overflow so I am not sure if it's possible.

You can try something like:
```java
InputStream inputStream = new InputStream() {
    final long SIZE = (long) (Integer.MAX_VALUE * 1.5);
    long read = 0;

    @Override
    public int available() {
        return (int) Math.min(SIZE - read, Integer.MAX_VALUE);
    }

    @Override
    public int read() {
        return (SIZE - read++ > 0) ? 1 : -1;
    }
};
```

prrace commented:
It isn't clear to me that this is "better".
Before although it would jump to zero it would continue to increment and so the user would know something is happening. Now it just gets to max and looks "stuck".
So I don't think it should be fixed this way.

If we could find a way to use longs here .. but then there are other problems .. 
The ProgressMonitor usage seems a bit less than ideal to me.
This class bases its estimate of max to pass to ProgressMonitor on InputStream.available().
If it is a real file, then this may be fine .. but available() returns an int .. so I suppose can't tell you the real size of a large file.
If it is any other kind of stream, then the size estimate is perhaps worse than useless.
It must be continually re-evaluating max in such a case ? Right ?

Perhaps that tells us what we should do 

Rather than sitting at max, or jumping back to zero, we can go back to something like
90% .. or 50% .. and so we are signalling we've had to re-evaluate where we are but ar still reading.

BTW I don't see why you need a large file for a test. You can just have an InputStream subclass which keeps generating data ..

prsadhuk commented:
> Before although it would jump to zero it would continue to increment and so the user would know something is happening. Now it just gets to max and looks "stuck".

Actually, the progressbar will be **blank** as ProgressMonitorInputStream will call `setProgress(nread+=nr)` ie with -ve value due to integer overflow so progress bar will not be 0 but somewhat (-)2147483125 so **we will not see any increment** also in the progressbar till it reaches 0 from such -ve value

I guess it will be weird to jump back from 99% to ~50% and start again in the same progress monitor (not sure if it's possible also as we support int and not long)
so
I modified the fix to recalculate the available size after MAX_INT is read and close the current progress monitor and recreate another progress monitor with recalculated max and show the progress from MAX_INT onwards to EOF.

> BTW I don't see why you need a large file for a test. You can just have an InputStream subclass which keeps generating data ..

I have added a test.

prrace commented on src/java.desktop/share/classes/javax/swing/ProgressMonitorInputStream.java:
```diff
@@ -88,6 +89,8 @@ public ProgressMonitorInputStream(Component parentComponent,
             size = 0;
         }
+        parentComponent = parentComponent;
+        message = message;
```
These are both wrong I am sure. You want to assign to this.<var> don't you ? Not update the local var.

prsadhuk replied:
Yes. Thanks for pointing out. Updated..

prsadhuk commented:
/integrate
